Initial Parse trait PR

kangalio opened the pull request:
This PR adds a new `serenity::utils::Parse` trait, mainly for https://github.com/serenity-rs/framework/issues/24.

As described in the docstring below, the `Parse` trait works like `std::str::FromStr`, but it also supports Discord model types like `Member` or `Message`.
```rust
/// Parse a value from a string in context of a received message.
///
/// This trait is a superset of [`std::str::FromStr`] (the trait used in `str::parse`). The
/// difference is that this trait supports serenity-specific Discord types like [`Member`], [`Role`],
/// or [`Emoji`].
///
/// Trait implementations may do network requests as part of their parsing procedure.
///
/// Useful for implementing command argument parsing frameworks.
#[async_trait::async_trait]
pub trait Parse: Sized {
    /// The associated error which can be returned from parsing.
    type Err;

    /// Parses a string `s` as a command parameter of this type.
    async fn parse(ctx: &Context, msg: &Message, s: &str) -> Result<Self, Self::Err>;
}
```

In this PR, support for `Member` and `Message` is implemented.

In a later PR, I hope to add support for more types:
- `User`/`UserId`
- `Channel`/`ChannelId`
- `GuildId`
- `ChannelCategory`
- `Role`/`RoleId`
- `Invite`
- `Emoji`/`EmojiId`/`EmojiIdentifier`
- `Colour`
- did I miss one?

kangalio committed:
Add Parse trait with two implementations

Implemented for Member and Message (and for all T where T: FromStr)

kangalio committed:
Fix formatting and make compatible with features

kangalio committed:
Small tweaks

- Improve documentation
- Only provide Parse impl for Member when cache feature is enabled
- Minimal code improvement

arqunis commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,186 @@
+            return Ok(msg);
+        }
+
+        if cfg!(feature = "gateway") {
```
The `gateway` feature implies the `http` feature, but I don't see why we couldn't depend on `http` directly. 
```suggestion
        if cfg!(feature = "http") {
```

kangalio replied:
That's just my unfamiliarity with serenity's features (I just use default features for my own bots :sweat_smile:) Thanks for the correction

kangalio committed:
Gate on http feature instead of gateway

kangalio commented:
Do you know how to stop GitHub from showing the "Changes requested" warning/info? The requested changes have been resolved, what else does it want

arqunis commented:
It has to be approved by the reviewer for it to go away.

Lakelezz commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,186 @@
+impl Parse for Member {
+    type Err = MemberParseError;
+
+    #[doc = ""] // to get impl documentation to show
```
Comments go before the related code item.
```suggestion
    // to get impl documentation to show
    #[doc = ""] 
```

Lakelezz commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,186 @@
+pub enum MessageParseError {
+    Malformed,
+    Http(SerenityError),
+    /// When the `gateway` feature is disabled and the required information was not in cache
```
```suggestion
    /// When the `gateway` feature is disabled and the required information was not in cache.
```

Lakelezz commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,186 @@
+            Self::Malformed => {
+                write!(f, "provided string did not adhere to any known guild message format")
+            },
+            Self::Http(e) => write!(f, "failed to request message data via HTTP: {}", e),
```
```suggestion
            Self::Http(e) => write!(f, "Failed to request message data via HTTP: {}.", e),
```

kangalio replied:
I see, I wasn't sure which casing should be preferred for error messages inside serenity.

It may be of interest that even existing code isn't fully consistent about this: some code uses uppercase errors, some lowercase:
![Screenshot_20210223_201001](https://user-images.githubusercontent.com/21220820/108895009-50fac500-7613-11eb-8e60-dd5c8c6d60b3.png)

kangalio replied:
Another thing: with the convention being to add a trailing full stop to error messages, wouldn't `{}.` expand to two full stops? Think `Failed to request message data via HTTP: Some HTTP specific error..` (note the `..`)

Lakelezz commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,186 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Malformed => {
+                write!(f, "provided string did not adhere to any known guild message format")
```
```suggestion
                write!(f, "Provided string did not adhere to any known guild message format.")
```

Lakelezz commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,186 @@
+impl Parse for Message {
+    type Err = MessageParseError;
+
+    #[doc = ""] // to get impl documentation to show
```
```suggestion
    // to get impl documentation to show
    #[doc = ""] 
```

Lakelezz commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,186 @@
+
+// Questions:
+// - Should the error enums be #[non_exhaustive]?
+// - Should the MessageParseError implement source()? It may be too much boilerplate (maybe add thiserror dependency?)
```
This does not belong in final code, but should be discussed before merging this.

Lakelezz commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,186 @@
+
+/// Parse a value from a string in context of a received message.
+///
+/// This trait is a superset of [`std::str::FromStr`] (the trait used in [`str::parse`]). The
+/// difference is that this trait supports serenity-specific Discord types like [`Member`],
+/// [`Role`] (soon), or [`Emoji`] (soon).
```
If it does not support the mentioned items, they should not be listed here (see the *soon*), but items that it supports. Alternatively, generalise the sentence.

Lakelezz commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,186 @@
+            Self::Http(e) => write!(f, "failed to request message data via HTTP: {}", e),
+            Self::HttpNotAvailable => write!(
+                f,
+                "gateway feature is disabled and the required information was not in cache"
```
```suggestion
                "Gateway feature is disabled and the required information was not in cache."
```

Lakelezz reviewed (changes requested):
Thanks for your pull request.
I changed the punctuation and I do think the comments have a few questions that must be solved *before* merging.

kangalio commented:
Thank you for the review Lakelezz. As suggested, here are remaining questions to discuss (extracted from comment inside code):

- Should the error enums be #[non_exhaustive]?
- Should the MessageParseError implement source()? It may be too much boilerplate (maybe add thiserror dependency?)

kangalio committed:
Fix up some formatting

- Remove questions about PR from source code
- Don't mention features that aren't supported yet in docstring
- Move line comments to the previous line instead of being inline
- Add trailing period to docstring

kangalio committed:
Capitalize error messages and comments

arqunis commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,184 @@
+///
+/// Trait implementations may do network requests as part of their parsing procedure.
+///
+/// Useful for implementing command argument parsing frameworks.
```
```suggestion
/// Useful for implementing argument parsing in command frameworks.
```

arqunis commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,184 @@
+use crate::{model::prelude::*, prelude::*};
```
```suggestion
use crate::model::prelude::*;
use crate::prelude::*;
```

arqunis commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,184 @@
+/// Error that can be returned from [`Member::parse`].
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+pub enum MemberParseError {
+    GuildNotInCache,
+    NotFoundOrMalformed,
```
These variants should also be documented.

arqunis commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,184 @@
+impl Parse for Member {
+    type Err = MemberParseError;
+
+    // To get impl documentation to show.
+    #[doc = ""]
```
I don't find much reason to have to do this. I've tried building the documentation without this and by expanding the impl, I was able to read the above documentation just fine. 

```suggestion
```

arqunis commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,184 @@
+/// Error that can be returned from [`Message::parse`].
+#[derive(Debug)]
+pub enum MessageParseError {
+    Malformed,
+    Http(SerenityError),
```
These variants should be documented.

arqunis commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,184 @@
+impl Parse for Message {
+    type Err = MessageParseError;
+
+    // To get impl documentation to show.
+    #[doc = ""]
```
Ditto.
```suggestion
```

arqunis commented on src/utils/parse.rs:
```diff
@@ -0,0 +1,184 @@
+
+/// Parse a value from a string in context of a received message.
+///
+/// This trait is a superset of [`std::str::FromStr`] (the trait used in [`str::parse`]). The
```
Users can refer to `FromStr`'s documentation to discover where it is used, no need to explain it here.
```suggestion
/// This trait is a superset of [`std::str::FromStr`]. The
```

arqunis commented:
> Should the error enums be `#[non_exhaustive]`?

All public structs and enums should ideally be marked `#[non_exhaustive]`. It allows us to add new fields/variants without breaking the API.

> Should the `MessageParseError` implement `source()`? It may be too much boilerplate (maybe add `thiserror` dependency?)

If `source` can be implemented for error structs/enums, it should be. Also, while it could ease writing error structs/enums, I wouldn't add `thiserror` just for that. If we were writing them often, it could be beneficial, but I'd like to avoid increasing compilation times for the library.

kangalio committed:
Apply suggestions from code review

Co-authored-by: Alex M. M. <acdenissk69@gmail.com>

kangalio committed:
Add error variants documentation

kangalio committed:
Merge branch 'parse' of https://github.com/kangalioo/serenity into parse

kangalio committed:
Make Parse trait Error's non_exhaustive

kangalio commented:
Okay, I made the newly added error enums non_exhaustive. They also have a `source` implementation where appropriate (it already existed, not sure in which commit I added that).

Are there more things left to tweak?
